### PR TITLE
cellSaveData: Add autosave indicator

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
This adds standard autosave and autoload indicators that would appear on PS3 when the game requests them. Since RPCS3 saves basically instantaneously, this auto indicator simply appears for a few seconds to let the user known when the game autosaves.

### CODE REVIEW

**Comment by jeremybris** [comment]
Can you please add a feature where... when the autosave indicator appears and you go to close the emulator, it opens a prompt saying:

`"Closing while saving may cause damage to your save. Are you sure you want to close the emulator? - Yes / No"`

I like this feature in PCSX2. It's probably unnecessary since RPCS3 saving is so fast, but I would rather be safe than sorry, and it might even help prevent users reporting issues that are user error.

**Comment by elad335** [comment]
It's not relevant for RPCS3, our saves are superior.

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 1]
```diff
@@ -198,18 +198,29 @@ namespace rsx
...
 			for (auto it = vis_set.rbegin(); it != vis_set.rend(); ++it, ++index)
 			{
 				if (origin == message_pin_location::top) [[ likely ]]
 				{
-					it->update(index, cur_time, y_offset);
+					it->update(index, cur_time, x_offset, y_offset);
 					y_offset += (spacing + it->h);
 				}
-				else
+				else if (origin == message_pin_location::bottom)
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:210]: Use a switch

File: rpcs3/Emu/Cell/Modules/cellSaveData.cpp [hunk 2]
```diff
@@ -1750,6 +1751,48 @@ static NEVER_INLINE error_code savedata_op(ppu_thread& ppu, u32 operation, u32 v
...
+		case CELL_SAVEDATA_INDICATORPOS_LOWER_LEFT:
+			msg_location = rsx::overlays::message_pin_location::bottom;
+			break;
+		case CELL_SAVEDATA_INDICATORPOS_UPPER_RIGHT:
+			msg_location = rsx::overlays::message_pin_location::top_right;
+			break;
+		case CELL_SAVEDATA_INDICATORPOS_LOWER_RIGHT:
+			msg_location = rsx::overlays::message_pin_location::bottom_right;
+			break;
+		}
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellSaveData.cpp:1786]: Missing default

File: rpcs3/Emu/Cell/Modules/cellSaveData.cpp [hunk 3]
```diff
@@ -1750,6 +1751,48 @@ static NEVER_INLINE error_code savedata_op(ppu_thread& ppu, u32 operation, u32 v
...
+		if (operation == SAVEDATA_OP_AUTO_SAVE || operation == SAVEDATA_OP_LIST_AUTO_SAVE)
+		{
+			msg_text = localized_string_id::CELL_SAVEDATA_AUTOSAVE;
+		}
+		else if (operation == SAVEDATA_OP_AUTO_LOAD || operation == SAVEDATA_OP_LIST_AUTO_LOAD)
+		{
+			msg_text = localized_string_id::CELL_SAVEDATA_AUTOLOAD;
+		}
+
+		u32 indicator_pos = (statSet->indicator->dispPosition & 0x0F);
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellSaveData.cpp:1769]:
```suggestion
		const u32 indicator_pos = (statSet->indicator->dispPosition & 0x0F);
```

**Comment by elad335** [comment]
Was testing something / not intended for merge yet

**Comment by Megamouse** [comment]
Did you forget about the review changes @NicknineTheEagle  ?

**Comment by NicknineTheEagle** [comment]
> Did you forget about the review changes @NicknineTheEagle ?

Ah, sorry - no, I didn't forget, I was just procrastinating on this because I need to add UI element to toggle the feature and I really don't like working with anything UI.

**Comment by digant73** [comment]
if this PR is still valid on top of the current RPCS3, isn't it good to be merged and someone else could make this feature configurable on the GUI?

**Comment by digant73** [comment]
@NicknineTheEagle check your repo. I opened a PR to add the GUI support + the reviews proposed here

**Comment by NicknineTheEagle** [comment]
> @NicknineTheEagle check your repo. I opened a PR to add the GUI support + the reviews proposed here

Thank you.

File: rpcs3/Emu/RSX/Overlays/overlay_message.h [hunk 4]
```diff
@@ -55,29 +57,44 @@ namespace rsx
...
+					queue = &m_ready_queue_bottom_right;
+					break;
+				case message_pin_location::bottom_left:
+					queue = &m_ready_queue_bottom_left;
+					break;
+				case message_pin_location::top_right:
+					queue = &m_ready_queue_top_right;
+					break;
+				case message_pin_location::top_left:
+				default:
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.h:80]: Remove the default here. It's an enum class.

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 5]
```diff
@@ -283,10 +308,15 @@ namespace rsx
...
-			case message_pin_location::bottom:
-				return check_list(m_ready_queue_bottom) || check_list(m_visible_items_bottom);
+			case message_pin_location::bottom_right:
+				return check_list(m_ready_queue_bottom_right) || check_list(m_visible_items_bottom_right);
+			case message_pin_location::bottom_left:
+				return check_list(m_ready_queue_bottom_left) || check_list(m_visible_items_bottom_left);
+			case message_pin_location::top_right:
+				return check_list(m_ready_queue_top_right) || check_list(m_visible_items_top_right);
+			case message_pin_location::top_left:
+			default:
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:318]: Same here

File: rpcs3/Emu/RSX/Overlays/overlay_message.cpp [hunk 6]
```diff
@@ -197,19 +197,31 @@ namespace rsx
...
 					y_offset += (spacing + it->h);
-					it->update(index, cur_time, virtual_height - y_offset);
+					it->update(index, cur_time, x_offset, virtual_height - y_offset);
+					break;
+				case message_pin_location::top_right:
+					it->update(index, cur_time, virtual_width - x_offset - it->w, y_offset);
+					y_offset += (spacing + it->h);
+					break;
+				case message_pin_location::top_left:
+				default:
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Overlays/overlay_message.cpp:221]: Same here

File: rpcs3/Emu/Cell/Modules/cellSaveData.cpp [hunk 7]
```diff
@@ -1750,6 +1752,49 @@ static NEVER_INLINE error_code savedata_op(ppu_thread& ppu, u32 operation, u32 v
...
+			msg_location = rsx::overlays::message_pin_location::bottom_right;
+			break;
+		case CELL_SAVEDATA_INDICATORPOS_LOWER_LEFT:
+			msg_location = rsx::overlays::message_pin_location::bottom_left;
+			break;
+		case CELL_SAVEDATA_INDICATORPOS_UPPER_RIGHT:
+			msg_location = rsx::overlays::message_pin_location::top_right;
+			break;
+		case CELL_SAVEDATA_INDICATORPOS_UPPER_LEFT:
+		default:
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/cellSaveData.cpp:1785]: Maybe also add CELL_SAVEDATA_INDICATORPOS_CENTER here to make it more explicit that this wasn't forgotten.

**Pull request merged**